tests/tests_fabric/strategies/test_model_parallel_integration.py

```diff
@@ -0,0 +1,489 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from pathlib import Path
+from unittest import mock
+
+import pytest
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from lightning.fabric import Fabric
+from lightning.fabric.strategies import ModelParallelStrategy
+from lightning.fabric.utilities.load import _load_distributed_checkpoint
+from torch.utils.data import DataLoader, DistributedSampler
+
+from tests_fabric.helpers.datasets import RandomDataset
+from tests_fabric.helpers.runif import RunIf
+
+
+@RunIf(min_torch="2.3", standalone=True, min_cuda_gpus=4)
+def test_setup_device_mesh():
+    from torch.distributed.device_mesh import DeviceMesh
+
+    for dp_size, tp_size in ((1, 4), (4, 1), (2, 2)):
+        strategy = ModelParallelStrategy(
+            parallelize_fn=(lambda m, _: m),
+            data_parallel_size=dp_size,
+            tensor_parallel_size=tp_size,
+        )
+        fabric = Fabric(accelerator="auto", devices=4, strategy=strategy)
+        fabric.launch()
+
+        device_mesh = fabric.strategy.device_mesh
+        assert isinstance(device_mesh, DeviceMesh)
+        assert device_mesh.device_type == fabric.device.type
+        assert device_mesh.mesh_dim_names == ("data_parallel", "tensor_parallel")
+        assert device_mesh.size(0) == dp_size
+        assert device_mesh.size(1) == tp_size
+        assert device_mesh.ndim == 2
+
+        fabric.barrier()
+
+    # Passing sizes that don't multiply to the world size raises an error
+    for invalid_dp_size, invalid_tp_size in ((1, 1), (2, 3), (4, 2)):
+        strategy = ModelParallelStrategy(
+            parallelize_fn=(lambda m, _: m),
+            data_parallel_size=invalid_dp_size,
+            tensor_parallel_size=invalid_tp_size,
+        )
+        fabric = Fabric(accelerator="auto", devices=4, strategy=strategy)
+        with pytest.raises(RuntimeError, match="multiplied should equal the world size"):
+            fabric.launch()
+
+        fabric.barrier()
+
+    # Passing "auto" will select internode and intranode dimensions automatically
+    strategy = ModelParallelStrategy(
+        parallelize_fn=(lambda m, _: m),
+        data_parallel_size="auto",
+        tensor_parallel_size="auto",
+    )
+    fabric = Fabric(accelerator="auto", devices=4, num_nodes=1, strategy=strategy)
+    fabric.launch()
+    assert fabric.strategy.device_mesh.mesh_dim_names == ("data_parallel", "tensor_parallel")
+    assert fabric.strategy.device_mesh.size(0) == 1
+    assert fabric.strategy.device_mesh.size(1) == 4
```

Comment:
CI runs with 2 devices. This could introduce a deadlock if PyTorch adds a collective call in the device mesh. Could we change it to 2 just in case?

Reply:
For 2D I need at least 2*2=4 devices. 
What do you mean this could add a deadlock. Where? I'm calling it on all ranks.

Reply:
I didn't notice that you had `min_cuda_gpus=4`.

Be aware then that this will not run on our CI (gets skipped) because all our agents run with only 2 visible CUDA devices.

https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=201341&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=8e4ceb7c-ceed-5ee0-b6fc-c9023e41cb74&l=1306

Reply:
Yes I'm aware. I was torn between doing this vs. simulating it on CPU. DTensor works on CPU, but I'd still need processes spawned for proper e2e testing and we currently don't have the combination of standalone=True and min_gpus=x in the CI. Also I wasn't convinced enough that CPU testing would be representative enough. Maybe we need to revisit this later.